pkg/util/clusterregistry/external.go

```diff
@@ -0,0 +1,69 @@
+package clusterregistry
+
+import (
+	"time"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clientcmd"
+	cr_v1a1 "k8s.io/cluster-registry/pkg/apis/clusterregistry/v1alpha1"
+	clusters "k8s.io/cluster-registry/pkg/client/clientset/versioned/typed/clusterregistry/v1alpha1"
+)
+
+type ExternalClusterRegistry struct {
+	namespace    string
+	rest         *clusters.ClusterregistryV1alpha1Client
+	cache        cache.ThreadSafeStore
+	lastRun      time.Time
+	resyncPeriod time.Duration
+}
+
+func NewExternalClusterRegistry(url string, resyncPeriod time.Duration) (*ExternalClusterRegistry, error) {
+	cfg, err := clientcmd.BuildConfigFromFlags(url, "")
+	if err != nil {
+		return nil, err
+	}
+
+	rest, err := clusters.NewForConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ExternalClusterRegistry{
+		namespace:    "kube-system",
+		rest:         rest,
+		cache:        cache.NewThreadSafeStore(cache.Indexers{}, cache.Indices{}),
+		resyncPeriod: resyncPeriod,
+	}, nil
+
+}
+
+func (e *ExternalClusterRegistry) GetClusters() ([]*cr_v1a1.Cluster, error) {
+	if time.Since(e.lastRun) > e.resyncPeriod {
```

Comment:
- `ExternalClusterRegistry` is supposed to be thread-safe, right? Currently it is not because `e.lastRun` is read and written without any synchronization means.
- Should likely avoid making multiple concurrent calls to the registry because why if you can do a single call and then serve data to multiple callers.
- Since you are maintaining a cache here, why not make it asynchronous and decouple reads from cache refreshes? Have a goroutine that is periodically refreshing the cache and always serve data from it? Can even use watch if it is supported to refresh data in the cache. How sensitive are we to stale information in that cache?